### PR TITLE
Correction for is_equal function

### PR DESCRIPTION
**Pull request by khancyr**
The current solution doesn't works for int as epsilon for int is 0 . But as fabsf always return a float and this function is only used with float and not double, we should use FLT_EPSILON in all case

### CODE REVIEW

**Comment by khancyr** [comment]
other solutions are : 
- use <= for comparison and then change the test EXPECT_FALSE(is_equal(1.f,  1.f + FLT_EPSILON)); to EXPECT_TRUE(is_equal(1.f,  1.f + FLT_EPSILON));
- have a different return generate a compile time for int and float by using something like std::enable_if , I try but failed ...

**Comment by OXINARF** [comment]
I don't think this is the correct way to fix is_equal.

> using something like std::enable_if , I try but failed ...

Why did it failed? You couldn't code it in a way that it worked or is the support for enable_if missing?

**Comment by khancyr** [comment]
I failed because I am not used to template and less to the use of enable_if

**Comment by OXINARF** [comment]
Ok, I've requested @lucasdemarchi to review, let's wait for his opinion on this.

**Comment by kwikius** [comment]
The following appears to work, if you change the comparison to less than or equal and also removes the issues about using other than float. Seems to pss basic tests

```c++
#include <cmath>
#include <type_traits>
#include <limits>
#include <cstdint>

template <class ArithmeticL, class ArithmeticR>
bool is_equal(const ArithmeticL v_1, const ArithmeticR v_2)
{
    static_assert(std::is_arithmetic<ArithmeticL>::value, "template parameter not of type float or int");
    static_assert(std::is_arithmetic<ArithmeticR>::value, "template parameter not of type float or int");
    return std::abs(v_1 - v_2) <= std::numeric_limits<decltype(v_1 - v_2)>::epsilon();
}

#include <iostream>

/*
  do some tests
*/
template <class TL, class TR>
void do_equal(TL const & lhs, TR const & rhs)
{
    if (is_equal(lhs,rhs)){
      std::cout << lhs << " == " << rhs << '\n';
    }else{
      std::cout << lhs << " != " << rhs << '\n';
    }
    
}

int main()
{
   int v1 = 1;
   long v2 = 2L;

   float v3 = 1.f;
   double v4 = 2.;
   unsigned int v5 = 'X';
   uint8_t  v6 = 'X';

   do_equal(v1,v2);
   do_equal(v1,v3);
   do_equal(v1,v4);
   do_equal(v1,v4);
   do_equal(v1,v5);
   do_equal(v1,v6);

   do_equal(v2,v2);
   do_equal(v2,v3);
   do_equal(v2,v4);
   do_equal(v2,v4);
   do_equal(v2,v5);
   do_equal(v2,v6);

   do_equal(v3,v2);
   do_equal(v3,v3);
   do_equal(v3,v4);
   do_equal(v3,v4);
   do_equal(v3,v5);
   do_equal(v3,v6);

   do_equal(v4,v2);
   do_equal(v4,v4);
   do_equal(v4,v4);
   do_equal(v4,v4);
   do_equal(v4,v5);
   do_equal(v4,v6);

   do_equal(v5,v2);
   do_equal(v5,v5);
   do_equal(v5,v5);
   do_equal(v5,v5);
   do_equal(v5,v5);
   do_equal(v5,v6);

   do_equal(v6,v2);
   do_equal(v6,v6);
   do_equal(v6,v6);
   do_equal(v6,v6);
   do_equal(v6,v5);
   do_equal(v6,v6);
}
```

**Comment by khancyr** [comment]
@kwikius yes I agree but that means that we accept that 1.0 == 1.00001. i am not sure that the 1e-5 precision is needed for now or not

**Comment by khancyr** [comment]
What I am thinking is something like 
````
template <class FloatOne, class FloatTwo>
typename std::enable_if< (std::is_integral<FloatOne>::value ),  bool>::type is_equal(const FloatOne v_1, const FloatTwo v_2)
{
    return v_1 == v_2;
}

template <class FloatOne, class FloatTwo>
typename std::enable_if< (std::is_floating_point<FloatOne>::value), bool>::type  is_equal(const FloatOne v_1, const FloatTwo v_2)
{
    return fabsf(v_1 - v_2) < std::numeric_limits<decltype(v_1 - v_2)>::epsilon();
}

````

But it is bothering me that it is template <class FloatOne, class FloatTwo> and I can only make std::enable_if< (std::is_floating_point<FloatOne>::value), bool>::type working, I cannot find a solution with FloatOne and FloatTwo inside enable_if ...

**Comment by kwikius** [comment]
@khancyr 
You can use [std::common_type](http://en.cppreference.com/w/cpp/types/common_type)<T...>::type to find the converted type of the result
I added the cast in the body of the integral version to kill warnings about comparisons of signed v unsigned. Alternatively it could just not be allowed.

I would suggest adding some tests to just confirm this is working as intended, 
https://github.com/ArduPilot/ardupilot/tree/master/tests
but I don't know where they should be added

```c++
template <class FloatOne, class FloatTwo>
typename std::enable_if< 
    std::is_integral< 
        typename std::common_type<FloatOne,FloatTwo>::type
    >::value
    ,bool
>::type 
is_equal(const FloatOne v_1, const FloatTwo v_2)
{
    typedef typename std::common_type<FloatOne,FloatTwo>::type common_type;
    return static_cast<common_type>(v_1) == static_cast<common_type>(v_2);
}

template <class FloatOne, class FloatTwo>
typename std::enable_if< 
    std::is_floating_point<
        typename std::common_type<FloatOne,FloatTwo>::type
    >::value
    , bool
>::type  
is_equal(const FloatOne v_1, const FloatTwo v_2)
{
    return std::abs(v_1 - v_2) < std::numeric_limits<decltype(v_1 - v_2)>::epsilon();
}
```

**Comment by khancyr** [comment]
@kwikius 
Whouaa ! Nice ! I already add some test in this PR but you can add more, it is in https://github.com/ArduPilot/ardupilot/tree/master/libraries/AP_Math/tests and use GTEST framework, to see result... I don't know, I use waf tests check-all and it works ^^
Normally we shouldn't be able to get unsigned comparison warning as is_equal is only define for int, short, float, and double!

Test pass with your solution that I think is the best

**Comment by kwikius** [comment]
Edit.
I updated the example to change the template Parameter names to 'Arithmetic'

. One day C++ will have concepts and then this naming will become a fully fledged Concept name
https://isocpp.org/blog/2016/02/a-bit-of-background-for-concepts-and-cpp17-bjarne-stroustrup

**Comment by khancyr** [comment]
Rebased with kwikius solution. maybe it is overkilling feature as in fact we only define the template to be use when both parameter are the same .

**Comment by kwikius** [comment]
Yes. To be honest, this would be better(faster, shorter code, better optimise) as an inline function in the header. Then there would be no need for explicit instantiations in the source file

**Comment by khancyr** [comment]
@kwikius feel free to open a PR !

File: libraries/AP_Math/AP_Math.cpp [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
-template <class FloatOne, class FloatTwo>
-bool is_equal(const FloatOne v_1, const FloatTwo v_2)
+template <class Arithmetic1, class Arithmetic2>
+typename std::enable_if<std::is_integral<typename std::common_type<Arithmetic1, Arithmetic2>::type>::value ,bool>::type
+is_equal(const Arithmetic1 v_1, const Arithmetic2 v_2)
 {
-    static_assert(std::is_arithmetic<FloatOne>::value, "template parameter not of type float or int");
-    static_assert(std::is_arithmetic<FloatTwo>::value, "template parameter not of type float or int");
+    static_assert(std::is_arithmetic<Arithmetic1>::value, "template parameter not of type int");
+    static_assert(std::is_arithmetic<Arithmetic2>::value, "template parameter not of type int");
```
Comment (lucasdemarchi) [inline libraries/AP_Math/AP_Math.cpp]: Do you really need the static asserts? The function shouldn't even be generated for other types.

**Comment by lucasdemarchi** [comment]
Other than my comment, LGTM

**Comment by khancyr** [comment]
@lucasdemarchi ok done !

**Comment by rmackay9** [comment]
I've tested this code with the Alt_Hold object avoidance (which is what led to finding the issue) and it works fine with the patches in this PR.  https://github.com/ArduPilot/ardupilot/pull/5410

So from my point of view, I'm happy for this to go into master.  We should probably squash some of the commits together but I guess either @khancyr could do that or whomever pulls it into master.

In any case, thanks!

**Comment by OXINARF** [comment]
Merged, thanks!

**Pull request merged**